cpp/src/parquet/column_writer.cc

```diff
@@ -768,194 +925,124 @@ void TypedColumnWriterImpl<DType>::CheckDictionarySizeLimit() {
-    values_to_write = num_levels;
-    spaced_values_to_write = num_levels;
-  }
-
-  // Not present for non-repeated fields
-  if (descr_->max_repetition_level() > 0) {
-    // A row could include more than one value
-    // Count the occasions where we start a new row
-    for (int64_t i = 0; i < num_levels; ++i) {
-      if (rep_levels[i] == 0) {
-        rows_written_++;
-      }
+    if (page_statistics_ != nullptr) {
+      const int64_t num_nulls = num_spaced_values - num_values;
+      page_statistics_->UpdateSpaced(values, valid_bits, valid_bits_offset, num_values,
+                                     num_nulls);
     }
-
-    WriteRepetitionLevels(num_levels, rep_levels);
-  } else {
-    // Each value is exactly one row
-    rows_written_ += static_cast<int>(num_levels);
   }
+};
 
-  if (descr_->schema_node()->is_optional()) {
-    WriteValuesSpaced(spaced_values_to_write, valid_bits, valid_bits_offset, values);
-  } else {
-    WriteValues(values_to_write, values);
-  }
-  *num_spaced_written = spaced_values_to_write;
+template <typename DType>
+Status TypedColumnWriterImpl<DType>::WriteArrowDictionary(const int16_t* def_levels,
+                                                          const int16_t* rep_levels,
+                                                          int64_t num_levels,
+                                                          const arrow::Array& array,
+                                                          ArrowWriteContext* ctx) {
+  // If this is the first time writing a DictionaryArray, then there's
+  // a few possible paths to take:
+  //
+  // - If dictionary encoding is not enabled, convert to densely
+  //   encoded and call WriteArrow
+  // - Dictionary encoding enabled
+  //   - If this is the first time this is called, then we call
+  //     PutDictionary into the encoder and then PutIndices on each
+  //     chunk. We store the dictionary that was written in
+  //     preserved_dictionary_ so that subsequent calls to this method
+  //     can make sure the dictionary has not changed
+  //   - On subsequent calls, we have to check whether the dictionary
+  //     has changed. If it has, then we trigger the varying
+  //     dictionary path and materialize each chunk and then call
+  //     WriteArrow with that
+  auto WriteDense = [&] {
+    std::shared_ptr<arrow::Array> dense_array;
+    RETURN_NOT_OK(MaterializeDictionary(array, properties_->memory_pool(), &dense_array));
+    return WriteArrowDense(def_levels, rep_levels, num_levels, *dense_array, ctx);
+  };
 
-  if (page_statistics_ != nullptr) {
-    page_statistics_->UpdateSpaced(values, valid_bits, valid_bits_offset, values_to_write,
-                                   spaced_values_to_write - values_to_write);
+  if (current_encoder_->encoding() != Encoding::PLAIN_DICTIONARY ||
+      !DictionaryDirectWriteSupported(array)) {
+    // No longer dictionary-encoding for whatever reason, maybe we
+    // never were or we decided to stop
```

Comment:
Technically you can feed this function a mix of DictionaryArray and non-DictionaryArray and it will encode them happily. I can add a comment clarifying